### PR TITLE
fix(typing): allow any object as first argument for get_type_hints

### LINKED ISSUE #4678: get_type_hints `obj` argument type seems too restrictive
The current type hints is:
```python
def get_type_hints(
    obj: Callable[..., Any], globalns: Optional[Dict[str, Any]] = ..., localns: Optional[Dict[str, Any]] = ...
) -> Dict[str, Any]: ...
```

The `Callable[..., Any]` seems really restrictive. The function works for example on an entire module, if you have a module such as:

```
# foo.py
VAR: str
```

The following code works fine and as expected:
```
import typing
from . import foo

typing.get_type_hints(foo)
# {'VAR': <class 'str'>}
```

However, there's a warning raised:
```
Argument 1 to "get_type_hints" has incompatible type Module; expected "Callable[..., Any]"
```

The fix seems easy, but I'm also wondering if there are other cases not covered by `Callable`. The function body seems to just call `getattr(obj, '__annotations__')` so I'm wondering if setting a type other than `Any` is even doable?

### PR DESCRIPTION
**Pull request by jd**
get_type_hints works on more object

Fixes #4678

### CODE REVIEW

**Comment by srittau** [comment]
I am not sure, why `object` has the `__annotations__` field, but that looks like an error:

```python
Python 3.9.0 (default, Oct  6 2020, 10:56:07) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> class Foo:
...     pass
... 
>>> Foo().__annotations__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Foo' object has no attribute '__annotations__'
>>> 
```

Similarly, using `get_type_hints()` on some objects fails:

```python
>>> from typing import get_type_hints
>>> get_type_hints(3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/srittau/.pyenv/versions/3.9.0/lib/python3.9/typing.py", line 1377, in get_type_hints
    raise TypeError('{!r} is not a module, class, method, '
TypeError: 3 is not a module, class, method, or function.
```

We need a different solution here. Maybe just `Any`, until we figure out something better.

**Comment by rohitkg98** [comment]
In the implementation, the allowed types have been mentioned as
```python
_allowed_types = (types.FunctionType, types.BuiltinFunctionType,
                  types.MethodType, types.ModuleType,
                  WrapperDescriptorType, MethodWrapperType, MethodDescriptorType)
```

we can use those in a Union.
refer [source](https://github.com/python/typing/blob/master/src/typing.py#L1467)

File: stdlib/3/typing.pyi [hunk 1]
```diff
@@ -585,17 +596,33 @@ class Pattern(Generic[AnyStr]):
...
+    MethodType,
+    ModuleType,
+)
+
+if sys.version >= (3, 7):
+    _get_type_hints_obj_allowed_types += (
+        WrapperDescriptorType,
+        MethodWrapperType,
+        MethodDescriptorType,
+    )
```
Comment (srittau) [inline stdlib/3/typing.pyi:611]: I doubt that works with all type checkers. Just have an `else` branch and list all types in all branches. Also, you need to use a `Union` here:

```python
if sys.version >= (3, 7):
    _get_type_hints_obj_allowed_types = Union[...]
else:
    _get_type_hints_obj_allowed_types = Union[...]
```

File: stdlib/3/typing.pyi [hunk 2]
```diff
@@ -585,17 +596,33 @@ class Pattern(Generic[AnyStr]):
...
+
 if sys.version_info >= (3, 9):
     def get_type_hints(
-        obj: Callable[..., Any],
+        obj: typing.Union[_get_type_hints_obj_allowed_types],
```
Comment (srittau) [inline stdlib/3/typing.pyi:615]: ... and then this `Union` (and the one below) are redundant.

**Review by srittau — DISMISSED** [review]

**Comment by jd** [comment]
I'm a bit worried about the Callable being refused now, I'm wondering how e.g. `FunctionType` and `Callable` are different?

**Comment by JelleZijlstra** [comment]
In theory, FunctionType is for concrete functions, and Callable is for anything that supports `__call__`. At least mypy, though, infers normal functions as Callables and doesn't really use FunctionType. You should make `get_type_hints()` accept Callable; this would give some false negatives around non-function callables but that seems better than rejecting them.

mypy-primer also finds some issues around Type. Since class objects can have `__annotations__`, perhaps we should support that too.

**Review by srittau — APPROVED** [review]
Thank for all your effort!

**Pull request merged**